### PR TITLE
Remove static lifetime from nonvolatile storage

### PR DESCRIPTION
**Pull request by lschuermann**
### Pull Request Overview

This pull request removes the `'static` lifetime requirements for the client and buffers in the `NonvolatileStorage` interface.

This relates to #1074.

### Testing Strategy

This pull request was tested by compiling.

### TODO or Help Wanted

All capsules using the `NonvolatileStorage` interface still need to be updated to use dynamic lifetimes. This will be done after the changes in `kernel/src/hil/nonvolatile_storage.rs` are approved.

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by phil-levis** [comment]
We've generally been allowing 'static to be explicit at the HIL level (e.g., SPI). Is there a reason to make it generic here? It seems like this is useful if we think there will be non-static implementations and so we want to keep the HIL generic. Do we think there will be (e.g., due to storage virtualization)? If there's a good use case that's not 'static then I'm all for this change.

**Comment by lschuermann** [comment]
Please excuse my delayed response. I've been writing some capsules using this HIL and it didn't feel good to be using `'static` here, as it then influenced other lifetime decisions.
One idea was - if and when something like #1291 would be merged - that the kernel could for instance initialize a nonvolatile storage, read some configuration and then deallocate it afterwards. When registering a static client and/or buffers, this seems impossible.

In summary, this was more of a *it may help in the future and shouldn't cause much trouble* PR. I'm fine with this being closed, and otherwise would invest the time to change all capsules to use appropriate lifetimes so that they would be non-static as well. (I'd be relying on the compiler for "testing", in case I don't have the hardware)

File: kernel/src/hil/nonvolatile_storage.rs [hunk 1]
```diff
@@ -4,29 +4,29 @@ use crate::returncode::ReturnCode;
 
 /// Simple interface for reading and writing nonvolatile memory. It is expected
 /// that drivers for nonvolatile memory would implement this trait.
-pub trait NonvolatileStorage {
-    fn set_client(&self, client: &'static NonvolatileStorageClient);
+pub trait NonvolatileStorage<'client, 'buffer: 'client> {
```
Comment (alevy) [inline kernel/src/hil/nonvolatile_storage.rs]: I don't there is a reason for these not to be the same lifetime. We should still be able to pass in a buffer with `&'static` lifetime, and it'll simply be upcast to whatever the `'client` liftetime is instantiated to

File: capsules/src/nonvolatile_to_pages.rs [hunk 2]
```diff
@@ -46,17 +46,17 @@ enum State {
     Write,
 }
 
-pub struct NonvolatileToPages<'a, F: hil::flash::Flash + 'static> {
+pub struct NonvolatileToPages<'driver, 'client, 'cbuf, F: hil::flash::Flash + 'static> {
```
Comment (alevy) [inline capsules/src/nonvolatile_to_pages.rs]: Ditto here: the lifetimes should be able to all be the same, no? The driver and client _definitely_ need always be equal (you'll never be able to instantiate something where the driver and client are not the same), and the buffer should always live _at least_ that long.

Comment (lschuermann) [reply capsules/src/nonvolatile_to_pages.rs]: I don't (wasn't required up to now to) have much experience with lifetimes, and yes, of course we can use upcasting here.
So having one lifetime `'a` should allow to use it like `'driver`, `'client: 'driver` and `'cbuf: 'client`, am I right?

**Comment by bradjc** [comment]
I added a commit that makes a single lifetime `'a` to match the other HILs.
